### PR TITLE
[Engine] allow (up/down)grade in create command

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR provides a basic API for ledger client to create contracts with down/up-grade support.
One can send a create Command where the payload contains 
 - (upgrade path) less fields than requires by the package definition (as long as the missing fields are optional), in such case the engine assume those fields are None
 - (downgrade path) more fields than requires by the package definition (additional field should be node), in such case the engine drop those fields. 

Here are some restrictions we assume for this first step: 
- For the sake of simplicity, we  forbid field reordering when upgrade is enable 
- The package of  explicit disclosed contract  is required 
- we ignore package ID complelty inside value (problably not what we want)

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 1]
```diff
@@ -58,7 +66,12 @@ private[lf] final class CommandPreprocessor(
       argument: Value,
   ): speedy.Command.Create = {
     discard(handleLookup(pkgInterface.lookupTemplate(templateId)))
-    val arg = valueTranslator.unsafeTranslateValue(Ast.TTyCon(templateId), argument)
+    val arg = valueTranslator.unsafeTranslateValue(
+      Ast.TTyCon(templateId),
+      argument,
+      allowFieldReordering = !enableContractUpgrading,
+      enableUpgrading = enableContractUpgrading,
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:73]: create argument can be (down/up)graded.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 2]
```diff
@@ -174,61 +182,146 @@ private[lf] final class ValueTranslator(
...
+                                  s"Mismatching record field label '$sourceField' (expecting '$targetField') for record $tyCon"
+                                )
+
+                          }
+                          val typ: Type = AstUtil.substitute(targetFieldType, subst)
+                          val sv: SValue = go(typ, v, newNesting)
+                          List(sv)
+                        case None => // DOWNGRADE
+                          // i ranges from 0 to numS-1. So i >= numT implies numS > numT
+                          assert((numS > i) && (i >= numT))
```
Comment (tudor-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:228]: This assertion is for documentation purposes I guess as it should be guaranteed externally by the `ImmArray` API

**Review by tudor-da — APPROVED** [review]
lgtm!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala [hunk 3]
```diff
@@ -49,13 +48,13 @@ private[lf] object Command {
...
       templateId: Identifier,
-      coid: SContractId,
+      coid: ContractId,
   ) extends Command
 
   /** Fetch a template, by interface */
   final case class FetchInterface(
       interfaceId: Identifier,
-      coid: SContractId,
+      coid: ContractId,
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:57]: It would have been nice to split cleanups like this into a separate PR. They cause a lot of noise, hiding the substantive changes, and unless I am misunderstanding something, are not essential to the upgrading feature.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 4]
```diff
@@ -174,61 +183,146 @@ private[lf] final class ValueTranslator(
...
+                  // - UPGRADE:   numT > numS : creates a None for each missing fields.
+                  // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
+                  //
+                  // When numS == numT, we wont hit the code marked either as UPGRADE or DOWNGRADE,
+                  // although it is still possible that the source and target types are different,
+                  // but since we don't consult the source type (may be unavailable), we wont know.
+
+                  val numS: Int = sourceElements.length
+                  val numT: Int = targetFieldsAndTypes.length
+
```
Comment (nickchapman-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:207]: Do we have plans to merge this code back with the code in `Speedy.scala` which it was cut-and-pasted from?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala [hunk 5]
```diff
@@ -243,10 +337,36 @@ private[lf] final class ValueTranslator(
   }
 
   // This does not try to pull missing packages, return an error instead.
-  def translateValue(
+  // TODO: https://github.com/digital-asset/daml/issues/17082
+  //  This is used by script and trigger, this should problaby use ValueTranslator.Config.Legacy
+  def strictTranslateValue(
```
Comment (nickchapman-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:342]: `strict` is not a great prefix here, since `Config.Legacy` has `allowFieldReordering = true`,
so is in some ways less strict. (I know you mean _strict_ as in _doesn't allow upgrade_)

Perhaps we should call it `legacyTranslateValue`.

Comment (nickchapman-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:342]: It is really the case that the legacy usage in `daml-script` etc requires the flexibility of field reordering?
Assuming so, do we have any plans to address this?

**Review by nickchapman-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:57]: I will revert that.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala [hunk 6]
```diff
@@ -280,10 +314,22 @@ class ValueTranslatorSpec
...
+          )
+        ) shouldBe a[Success[_]]
       )
       forEvery(testCasesForCid(illegalCid))((typ, value) =>
-        Try(valueTranslator.unsafeTranslateValue(typ, value)) shouldBe failure
+        Try(
+          valueTranslator.unsafeTranslateValue(
+            typ,
+            value,
+            Config.Legacy,
```
Comment (simonmaxen-da) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala:330]: Appreciate we need to support legacy translation but do we need tests to support the new non-legacy modes ?

**Review by simonmaxen-da — APPROVED** [review]
LGTM - with a couple of comments.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala]: 11097500c4fca0916d50b8f69b76f8a2c4a36cf7

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:212]: I have no concrete plan. If you want to factorize this code with the importValue, we probably have to try to factorize the whole function. I am not sure how to do that.
I suggest we revisit the question  when we have more advanced in the implementation of upgrade. 
Let's me add a TODO.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 7]
```diff
@@ -42,8 +110,10 @@ private[lf] final class CommandPreprocessor(
         )
       case _ =>
     }
-    val arg = valueTranslator.unsafeTranslateValue(Ast.TTyCon(disc.templateId), disc.argument)
-    valueTranslator.validateCid(disc.contractId)
+    // TODO: https://github.com/digital-asset/daml/issues/17082
+    //  for now we need the package of the disclosed contract
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:114]: It does not fully break downgradibility, it just require you have the two package vetted. 

>  Is there a issue specific to the support of downgraded disclosed contracts that can be created/referenced ?

The engine needs a package to make sens of data. It the package is unknown by the ledger, I think, it should be the most "upgraded" package the ledger know.  If the disclosed contract is compatible with this package we translate with this one, if it is not compatible we  crash (or ignore it) the contract as we will not be able to use it during interpretation. 

However currently there is no such API for the engine to get this package.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:214]: There is currently no concrete plan. 
Let's revisit the question when we are more advance with the upgrade implementation. 
I add a TODO.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:236]: I assume yes. 
This is a copy/paste of the code by @nickchapman-da  in Speedy.scala

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala:349]: > It is really the case that the legacy usage in daml-script etc requires the flexibility of field reordering?

This actually why I called it "strict" I assume daml-script/trigger does note require the flexibility. 

> Assuming so, do we have any plans to address this?

As soon we have a bit of time.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ValueTranslatorSpec.scala:330]: We need to adapt the tests that @nickchapman-da write for importValue
